azurelinuxagent/common/cgroupconfigurator.py

```diff
@@ -180,66 +248,86 @@ def __ensure_azure_slices_exist(self):
+            agent_drop_in_file_slice = "/lib/systemd/system/{0}.d/{1}".format(self._cgroups_api.get_agent_unit_name(), _AGENT_DROP_IN_FILE_SLICE)
+            agent_drop_in_file_cpu_accounting = "/lib/systemd/system/{0}.d/{1}".format(self._cgroups_api.get_agent_unit_name(), _AGENT_DROP_IN_FILE_CPU_ACCOUNTING)
+
+            files_to_create = []
+            return_azure_slice = False
 
             if os.path.exists(azure_slice):
-                # remove the overrides in case they were created by a previous version of the agent
-                self._cleanup_unit_file(azure_slice_override)
-                self._cleanup_unit_file(extensions_slice_override)
-                self._cleanup_unit_file(agent_drop_in_file)
-                return _AZURE_SLICE
-
-            if os.path.exists(azure_slice_override):
-                return _AZURE_SLICE
-
-            if not os.path.exists(azure_slice_override):
-                self._create_unit_file(azure_slice_override, _AZURE_SLICE_CONTENTS)
-                self._create_unit_file(extensions_slice_override, _EXTENSIONS_SLICE_CONTENTS)
-                drop_in_parent, _ = os.path.split(agent_drop_in_file)
-                if not os.path.exists(drop_in_parent):
-                    fileutil.mkdir(drop_in_parent, mode=0o755)
-                self._create_unit_file(agent_drop_in_file, _AGENT_DROP_IN_CONTENTS)
-                # reload the systemd configuration, but the new slice will not be used until the agent's service restarts
+                return_azure_slice = True
+            else:
+                files_to_create.append((azure_slice, _AZURE_SLICE_CONTENTS))
+
+            if not os.path.exists(vmextensions_slice):
+                files_to_create.append((vmextensions_slice, _VMEXTENSIONS_SLICE_CONTENTS))
+
+            if fileutil.findstr_in_file(agent_unit_file, "Slice=azure.slice"):
+                self._cleanup_unit_file(agent_drop_in_file_slice)
+            else:
+                if not os.path.exists(agent_drop_in_file_slice):
+                    files_to_create.append((agent_drop_in_file_slice, _AGENT_DROP_IN_FILE_SLICE_CONTENTS))
+
+            if fileutil.findstr_in_file(agent_unit_file, "CPUAccounting=yes"):
+                self._cleanup_unit_file(agent_drop_in_file_cpu_accounting)
+            else:
+                if not os.path.exists(agent_drop_in_file_cpu_accounting):
+                    files_to_create.append((agent_drop_in_file_cpu_accounting, _AGENT_DROP_IN_FILE_CPU_ACCOUNTING_CONTENTS))
+
+            if len(files_to_create) > 0:
                 try:
-                    shellutil.run_command(["systemctl", "daemon-reload"])
+                    for path, contents in files_to_create:
+                        self._create_unit_file(path, contents)
+
+                    # reload the systemd configuration, but the new slice will not be used until the agent's service restarts
```

Comment:
Would it be worthwhile to force a agent restart? Historically if its a long running VM the agent might not get restarted for a long long time. 

Maybe setup a cron job to restart the agent or something to avoid any changes in the daemon?

Reply:
At some point I considered that, but concluded it is not worth taking the risk of side effects (when is it a good time to restart the service? for example), plus better to avoid extra complications like managing the cron task.
The code already adapts to the agent running on either system or azure so functionality-wise it makes no difference.

Reply:
Yeah makes sense. Plus I suppose this is good for v1. If we feel the need to have some sort of retry we can probably address it later